components/tidb_query_datatype/src/codec/collation/encoding/utf8.rs

```diff
@@ -2,20 +2,37 @@
 
 use super::*;
 
-#[derive(Debug)]
-pub struct EncodingUTF8;
+pub trait UTF8CompatibleEncoding {
+    const NAME: &'static str;
+}
 
-impl Encoding for EncodingUTF8 {
+impl<T: UTF8CompatibleEncoding> Encoding for T {
     #[inline]
     fn decode(data: BytesRef) -> Result<Bytes> {
         match str::from_utf8(data) {
             Ok(v) => Ok(Bytes::from(v)),
-            Err(_) => Err(Error::cannot_convert_string("utf8mb4")),
+            Err(_) => Err(Error::cannot_convert_string(T::NAME)),
         }
     }
+}
```

Comment:
Why add this trait rather than impl `Encoding` on all of them?

Reply:
I think it can't save too much codes.

Reply:
`EncodingUTF8Mb4`, `EncodingUTF8` and `EncodingLatin1` share the same `decode` implemention, only different at the name. 

Sure it saves little codes, but it is not hard to understand and only have the file scope